src/batch/src/executor/join/lookup_join.rs

```diff
@@ -0,0 +1,592 @@
+/// 2) For each row R, find the rows that match with R on the join conditions on the probe side by
+/// using the `GatherExecutor` to query the probe side
+/// 3) Join R and the matching rows on the probe side together with either inner or left outer join
+/// 4) Repeat 2-3) for every row in the build side
+pub struct LookupJoinExecutor<GE> {
+    join_type: JoinType,
+    condition: Option<BoxedExpression>,
+    build_side_source: RowLevelIter,
+    build_side_data_types: Vec<DataType>,
+    probe_side_data_types: Vec<DataType>,
+    probe_side_source_params: ProbeSideSourceParams,
+    chunk_builder: DataChunkBuilder,
+    schema: Schema,
+    output_indices: Vec<usize>,
+    last_chunk: Option<SlicedDataChunk>,
+    identity: String,
+    unused: PhantomData<GE>, // Only exists so we can use generic type
+}
+
+impl<GE: 'static + GatherExecutor> Executor for LookupJoinExecutor<GE> {
+    fn schema(&self) -> &Schema {
+        &self.schema
+    }
+
+    fn identity(&self) -> &str {
+        &self.identity
+    }
+
+    fn execute(self: Box<Self>) -> BoxedDataChunkStream {
+        self.do_execute()
+    }
+}
+
+impl<GE: 'static + GatherExecutor> LookupJoinExecutor<GE> {
+    #[try_stream(boxed, ok = DataChunk, error = RwError)]
+    async fn do_execute(mut self: Box<Self>) {
+        self.build_side_source.load_data().await?;
+
+        loop {
+            let cur_row = self.build_side_source.get_current_row_ref();
+            if cur_row.is_none() {
+                break;
+            }
+            let cur_row = cur_row.unwrap();
+
+            let gather_executor = GE::new_boxed(&cur_row, &self.probe_side_source_params)?;
+            let mut probe_side_stream = gather_executor.execute();
+
+            let mut chunk_added = false;
+
+            loop {
+                let probe_side_chunk = probe_side_stream.next().await;
+                let chunk_is_none = probe_side_chunk.is_none();
+
+                // If None is received for the first time on the probe_side_stream AND it's a Left
+                // Outer Join AND no chunks have been added for the current row, don't break yet
+                // as we need to add NULL-padded values
+                if chunk_is_none && (self.join_type == JoinType::Inner || chunk_added) {
+                    break;
+                }
+
+                // Join the cur_row and the probe_side_chunk depending on the given join type
+                // Currently, Lookup Join only supports Inner and Left Outer Join
+                let join_result = match self.join_type {
+                    JoinType::Inner => self.do_inner_join(probe_side_chunk),
+                    // Only perform a Left Outer Join if chunk_is_none (and by extension no chunk
+                    // has been added yet for the current row)
+                    JoinType::LeftOuter if chunk_is_none => self.do_left_outer_join(),
+                    JoinType::LeftOuter if !chunk_is_none => self.do_inner_join(probe_side_chunk),
```

Comment:
In fact, we can also support `LeftAnti`, `LeftSemi` join.

Reply:
I think this can be added in a later PR